### PR TITLE
feat: MySQL & MSSQL query generator

### LINKED ISSUE #24466: [Task]:  MySQL and MSSQL query generator for one click binding
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

- Add support for mysql and ms sql query generators for one click binding
- [x] MySQL
- [x] MSSQL

### PR DESCRIPTION
**Pull request by berzerkeer**
## Description
- One click binding query generators for MySQL and MsSQL.

#### PR fixes following issue(s)
Fixes #24466 

#### Type of change
-  New feature (non-breaking change which adds functionality)

- [x] Manual
- [x] Jest
- [x] Cypress
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Test-plan-implementation#speedbreaker-features-to-consider-for-every-change) have been covered
- [x] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans/_edit#areas-of-interest)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [x] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

**Comment by berzerkeer** [comment]
Note : As of now the `mssql_spec.ts` for one click binding fails in cypress because of duplicate key issues on local, even when I delete stacks and restart servers. So running it in CI here.

Also as of now I believe a lot of these cypress JS commands are not converted to TS yet, if its already done. Please let me know , I will update the tests to use TS methods in a different follow up PR. Since that should be done for other one click binding specs aswell. 

cc: @Aishwarya-U-R  @NandanAnantharamu

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
+import oneClickBindingLocator from "../../../../../locators/OneClickBindingLocator";
+import * as _ from "../../../../../support/Objects/ObjectsCore";
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts:2]: Import only required objects from ObjectsCore.

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+const oneClickBinding = new OneClickBinding();
+
+describe("Table widget one click binding feature", () => {
+  it("should check that queries are created and bound to table widget properly", () => {
+    _.entityExplorer.DragDropWidgetNVerify(_.draggableWidgets.TABLE, 400);
+
+    _.entityExplorer.NavigateToSwitcher("Explorer");
+
+    _.dataSources.CreateDataSource("MsSql", true, true, "fakeapi");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts:13]: Refer to already existing way of using MsSQL ds & update this.

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
+import oneClickBindingLocator from "../../../../../locators/OneClickBindingLocator";
+import * as _ from "../../../../../support/Objects/ObjectsCore";
+import { OneClickBinding } from "../spec_utility";
+
+const oneClickBinding = new OneClickBinding();
+
+describe("Table widget one click binding feature", () => {
+  it("should check that queries are created and bound to table widget properly", () => {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts:8]: Index the `it` blocks.

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    _.agHelper.GetNClick(_.table._addNewRow, 0, true);
+
+    (cy as any).enterTableCellValue(0, 0, "S01E01");
+
+    (cy as any).enterTableCellValue(1, 0, "1");
+
+    (cy as any).enterTableCellValue(2, 0, " 1");
+
+    (cy as any).enterTableCellValue(3, 0, " 10");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts:61]: Use Ts helper method `EditTableCell` similar to other OneClickBinding folder specs that is fixed now.

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    _.assertHelper.AssertNetworkStatus("@postExecute");
+
+    _.agHelper.TypeText(_.table._searchInput, "Expanse");
+
+    _.assertHelper.AssertNetworkStatus("@postExecute");
+
+    _.agHelper.AssertElementExist(_.table._bodyCell("Expanse"));
+
+    (cy as any).wait(1000);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts:83]: Use` agHelper.Sleep();`

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts (hunk 4 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mssql_spec.ts:61]: Same comments for below spec too.

**Comment by berzerkeer** [comment]
@chandannkumar can you pick this one up next ? Thanks !

File: app/client/cypress/support/Pages/DataSources.ts [hunk 6]
```diff
@@ -975,6 +975,7 @@ export class DataSources {
       | "Oracle",
     navigateToCreateNewDs = true,
     testNSave = true,
+    databaseName?: string,
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:978]: Is this needed?

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 7]
```diff
@@ -1034,7 +1034,7 @@ export class AggregateHelper extends ReusableHelper {
     this.GetElement(selector)
       .closest("input")
       .scrollIntoView({ easing: "linear" })
-      .clear()
+      .clear({ force: true })
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/AggregateHelper.ts:1037]: Can you please get force as a parameter & use here? Not all will need force: true

File: app/client/cypress/e2e/Sanity/Datasources/MsSQL_Basic_Spec.ts [hunk 8]
```diff
@@ -128,6 +135,120 @@ describe("Validate MsSQL connection & basic querying with UI flows", () => {
     });
   });
 
+  it("3.One click binding - should check that queries are created and bound to table widget properly", () => {
+    entityExplorer.DragDropWidgetNVerify(draggableWidgets.TABLE, 450, 200);
+
+    entityExplorer.NavigateToSwitcher("Explorer");
+
+    entityExplorer.NavigateToSwitcher("Widgets");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Sanity/Datasources/MsSQL_Basic_Spec.ts:143]: Drag dropping table will automatically open Table's properties. Why we need these 2 lines?

File: app/client/cypress/e2e/Sanity/Datasources/MsSQL_Basic_Spec.ts [hunk 9]
```diff
@@ -128,6 +135,120 @@ describe("Validate MsSQL connection & basic querying with UI flows", () => {
...
+
+    entityExplorer.SelectEntityByName("Table1", "Widgets");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Sanity/Datasources/MsSQL_Basic_Spec.ts:145]: Also not needed.

File: app/client/cypress/e2e/Sanity/Datasources/MsSQL_Basic_Spec.ts [hunk 10]
```diff
@@ -128,6 +135,120 @@ describe("Validate MsSQL connection & basic querying with UI flows", () => {
...
+
+    agHelper.TypeText(table._searchInput, "Expanse");
+
+    assertHelper.AssertNetworkStatus("@postExecute");
+
+    agHelper.AssertElementExist(table._bodyCell("Expanse"));
+
+    agHelper.Sleep(1000);
+
+    // (cy as any).editTableCell(1, 0);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Sanity/Datasources/MsSQL_Basic_Spec.ts:213]: Can this comment be removed? As this method is not used anyways.

**Review by Aishwarya-U-R — CHANGES_REQUESTED** [review]

File: app/client/src/WidgetQueryGenerators/types.ts [hunk 11]
```diff
@@ -64,6 +64,14 @@ export type ActionConfigurationPostgreSQL = {
   pluginSpecifiedTemplates: Array<object>;
 };
 
+export type ActionConfigurationMySQL = {
+  pluginSpecifiedTemplates: Array<object>;
+};
+
+export type ActionConfigurationMSSQL = {
+  pluginSpecifiedTemplates: Array<object>;
```
Comment (sbalaji1192) [inline app/client/src/WidgetQueryGenerators/types.ts:72]: can we reuse ActionConfigurationPostgreSQL?

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 12]
```diff
@@ -1030,11 +1030,11 @@ export class AggregateHelper extends ReusableHelper {
     this.Sleep(500); //for value set to settle
   }
 
-  public UpdateInputValue(selector: string, value: string) {
+  public UpdateInputValue(selector: string, value: string, force = false) {
     this.GetElement(selector)
       .closest("input")
       .scrollIntoView({ easing: "linear" })
-      .clear()
+      .clear({ force })
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/AggregateHelper.ts:1037]: Wrong syntax , it should be { force: force }. If you are sending false again here - why is this even needed?

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mysql_spec.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+} from "../../../../../support/Objects/ObjectsCore";
+import { OneClickBinding } from "../spec_utility";
+
+const oneClickBinding = new OneClickBinding();
+
+describe("Table widget one click binding feature", () => {
+  it("1.should check that queries are created and bound to table widget properly", () => {
+    entityExplorer.DragDropWidgetNVerify(draggableWidgets.TABLE, 400);
+
+    dataSources.CreateDataSource("MySql");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mysql_spec.ts:18]: Is this tested after removing this  `entityExplorer.NavigateToSwitcher("Explorer");`? I think this is needed. Cross-check again.

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 12 above)
Comment (berzerkeer) [reply app/client/cypress/support/Pages/AggregateHelper.ts:1037]: Oh its not wrong syntax, in ES6 , if the key and value variable name is same ,  i.e `{ force : force }` doing `{force} is enough.

**Review by Aishwarya-U-R — COMMENTED** [review]
Approving Cypress changes!

File: app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mysql_spec.ts (hunk 13 above)
Comment (berzerkeer) [reply app/client/cypress/e2e/Regression/ClientSide/OneClickBinding/TableWidget/mysql_spec.ts:18]: Yup tested it passed.

**Comment by berzerkeer** [comment]
> Approving Cypress changes!

Thank you 💯

File: app/client/src/WidgetQueryGenerators/types.ts [hunk 14]
```diff
@@ -64,6 +64,10 @@ export type ActionConfigurationPostgreSQL = {
   pluginSpecifiedTemplates: Array<object>;
 };
 
+export type ActionConfigurationMySQL = ActionConfigurationPostgreSQL;
```
Comment (sbalaji1192) [inline app/client/src/WidgetQueryGenerators/types.ts:67]: we can rename ActionConfigurationPostgreSQL to ActionConfigurationSQL and use it.

Comment (berzerkeer) [reply app/client/src/WidgetQueryGenerators/types.ts]: done

**Review by sbalaji1192 — DISMISSED** [review]

**Comment by chandannkumar** [comment]
Observation @berzerkeer 
- Able to Insert, update in MySQL even with Read-only access in DB
https://www.loom.com/share/5547aaff75b6453885e66764549b7d87?sid=2ded1bf7-1d10-4197-bd3e-735b706a07d5

**Comment by berzerkeer** [comment]
> Observation @berzerkeer
> 
> * Able to Insert, update in MySQL even with Read-only access in DB
>   [loom.com/share/5547aaff75b6453885e66764549b7d87?sid=2ded1bf7-1d10-4197-bd3e-735b706a07d5](https://www.loom.com/share/5547aaff75b6453885e66764549b7d87?sid=2ded1bf7-1d10-4197-bd3e-735b706a07d5)

Does this happen with other Datasources too ? like postgres and Mongo ?

**Review by Aishwarya-U-R — APPROVED** [review]

**Comment by chandannkumar** [comment]
> Does this happen with other Datasources too ? like postgres and Mongo ?

For eg in Postgres when only Read access is configured, we can observe there is no editable field checked and no insert, update query created automatically after one click binding

**Comment by berzerkeer** [comment]
> > Does this happen with other Datasources too ? like postgres and Mongo ?
> 
> For eg in Postgres when only Read access is configured, we can observe there is no editable field checked and no insert, update query created automatically after one click binding

No, it does get created with postgres. 

Also I created a `READ_ONLY` Mysql datasource, and did `Generate page from CRUD` with that datasource. But I can still Update and Insert rows to that `READ_ONLY` datasource. 

Attached a video for reference. 
https://vdqm24wed6.vmaker.com/record/OLsIPiKIyAIl02jq

**Comment by chandannkumar** [comment]
Agree @berzerkeer, Above callout is an existing behavior.

QQ - Should we not restrict or inform users at the first place saying Datasource has only Read access and hence insert, update cannot be done
cc: @dilippitchika @sbalaji1192

**Comment by laveena-en** [comment]
Verified MSSQL for one-click binding. 
Apart from the existing issues stated [here](https://theappsmith.slack.com/archives/C02JV8G1MP0/p1687957379515339?thread_ts=1687526688.260129&cid=C02JV8G1MP0), this PR LGTM. (Props bound to `Total record count query would fail)

cc: @sbalaji1192 @chandannkumar

**Review by Aishwarya-U-R — APPROVED** [review]

**Comment by chandannkumar** [comment]
Testing is done for Snowflake DB support with One click binding and it's LGTM

**Comment by berzerkeer** [comment]
Created a [issue](https://github.com/appsmithorg/appsmith/issues/25149) to handle cypress failures.

**Pull request merged**